### DISCUSSION TITLE
[Bug] StrictMode combined with Controllers and async setValue's not working as expected

### DISCUSSION THREAD

**Discussion by LuisRizo** [react-hook-form/react-hook-form#6825] [Q&A] [answered]
### Version Number

7.17.1

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-js-forked-bldov?file=/src/App.js

### Steps to reproduce

1. Go to https://codesandbox.io/s/react-hook-form-js-forked-bldov?file=/src/App.js
2. Click on 'Start async setValue'
3. Notice that the output of `watch('color')` is still undefined (the Dropdown is still unset).

At first, I thought this was an issue with react-hook-form given that this bug does not occur on 7.11.1. But when I tried to recreate a Codesandbox example, I was not able to. It took me a long time to realize it was an issue related to React.StrictMode that we use in our app.

The conditions of the bug are the following:
1. Use React.StrictMode
2. Set the default value of a field to a value that is not `undefined`. Example: `''`. [This matters because of this code](https://github.com/react-hook-form/react-hook-form/pull/6700/files#diff-5a85c1643c71f45e050ce584d7c0db2bdf2e56f8a7dae86d74fedf01be2d2cabR939).
3. Invoke `setValue` outside of a click handler (promises can achieve this).

I understand that this bug is not 100% caused by `react-hook-form` (though it could be accounted for and fixed, as it was not occuring on 7.11.1), but I thought I should file this issue anyways as to prevent hours of debugging for other users of `React.StrictMode` and `react-hook-form` that might hit this edge-case.

### Expected behaviour

After calling `setValue('color', 'ocean')` in an asynchronous way, the form's state for that field should change appropriately (to `'ocean'` in this case).

### Workaround

To work around this bug, one can either disable StrictMode (as it doesn't affect production anyways), or we can set the default value to `undefined` (if possible). Both of these changes fix the bug. Another fix would be to call `setValue` inside the click handler in a synchronous way, but that is not possible for me.

In the Codesandbox, you can switch try switching the following commented-out code to see the workarounds:

![image](https://user-images.githubusercontent.com/25105030/137019354-d03747de-b0b6-435a-a6eb-5db33b6e6430.png)

You can switch one of them at a time

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
[App] color "" 
[Child] value "" 
[App] color "" 
[Child] value ""
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by LuisRizo** [comment]
My theory is that on strict mode, after the observers callback gets called
![image](https://user-images.githubusercontent.com/25105030/137019704-7e8dcd41-c748-4b10-9b18-1374e81ebe09.png)

there happens an extra re-render where the `value` is still the original, `defaultValue`. In this new re-render, `register` calls `set` function if the `defaultValue !== undefined`

![image](https://user-images.githubusercontent.com/25105030/137019951-fd117ef2-6299-438f-b856-7dff93aacca9.png)

So this `set` call now takes priority in the final re-render, and we end up never updating `value` inside `useController`.

As for possible solutions could be the following:
1. Also check for empty strings (along with `undefined`) before calling `set()` inside `register`.
2. Compare compare this new `value` with the `defaultValue` to avoid calling it multiple times inside `register`.
3. Undoing the refactor that caused this bug (I believe it could've been https://github.com/react-hook-form/react-hook-form/pull/5958).
4. Warn against using `React.StrictMode` on the README.md
5. Don't call `set` inside `register`. Move it to a useEffect inside `useForm` perhaps?

**Comment by bluebill1049** [comment, marked as answer]
thanks for the detailed investigation, one thing i sport is that we do not support `undefined` as `defaultValue`.

> The defaultValues for inputs are used as the initial value when a component is first rendered, before a user interacts with it. It is encouraged that you set defaultValues for all inputs to non-undefined values such as the empty string or null.

<img width="958" alt="Screen Shot 2021-10-13 at 8 57 32 am" src="https://user-images.githubusercontent.com/10513364/137034624-2f3d49be-5102-481e-9721-4d54022eb999.png">

**Comment by LuisRizo** [comment]
No worries and thanks for the reply!

Thanks also for pointing the docs. It's unfortunate that `undefined` is not a valid solution. That's the temporary solution we're using since we use `yup` for schema validation, which on that layer sets `undefined` values to empty strings. Removing StrictMode, which is the other option, isn't something we want to remove quite yet.

**Comment by bluebill1049** [comment]
`undefined` is going to be an issue with other hooks in this lib as well, because we use `undefined` to determine when to use `defaultValue` vs `defaultValues`. The main reason behind this is that we support uncontrolled input (native inputs).

**Comment by Joe312341** [comment]
Hi Bill, 

Was trying to integrate react-hook-form with google autocomplete component today and am facing the exact same issue.
If I do not use strict mode everything works fine.

Added a code sandbox but it needs a valid api key to work: https://codesandbox.io/s/adoring-feynman-el72i

In strict mode `setValue("streetAddress1", "unrelated string")` does not work and the field value never gets changed. Without strict mode everything works as expected.

Any suggestion to work around this without removing strict mode?

Thank you

**Comment by bluebill1049** [comment]
Thanks, @Joe312341 I would recommend producing the issue in a regular input or controlled component in which code is exposed, an external lib component is hard to investigate the issue. As the original issue reported is by design (limitation) of this lib.

**Comment by LuisRizo** [comment]
@Joe312341 I'm actually having this issue because we implemented this Google Autocomplete component, which we have the State and Country as a Dropdown. So we store the string from Google Autocomplete in react-hook-form, and render it as a Dropdown using Controlled components. However, this issue I was able to reproduce without having to add this third-party lib (react-select is only added for an example of a Controlled input).

The only difference I can tell from your example is that you were able to reproduce it without calling `setValue` in an asynchronous way. We use `geocodeByAddress` (from the same lib) in the selection handler, so we have to use Promises for calling `setValue`.

> undefined is going to be an issue with other hooks in this lib as well, because we use undefined to determine when to use defaultValue vs defaultValues. The main reason behind this is that we support uncontrolled input (native inputs).

@bluebill1049 So it's working for me because I'm not using `defaultValue`, just `defaultValues`?

What do you suggest to be an alternative solution to using `undefined` as the value for `defaultValues` or turning off StrictMode?

Personally, I feel like having `register()` call `set` every re-render when `defaultValue !== undefined` is something that can be improved upon. I'd suggest using hooks but `register` itself isn't a hook. But I'm guessing it can be moved somewhere else. What do you think?

**Comment by bluebill1049** [comment]
> I'd suggest using hooks but register itself isn't a hook. But I'm guessing it can be moved somewhere else. What do you think?

I would use null instead, plus native input can't support `undefined` as well and `register` is mostly made for native inputs. 

> So it's working for me because I'm not using defaultValue, just defaultValues?

what do you mean by that?
